proving_system/stark/src/air/example/cairo.rs

```diff
@@ -195,34 +488,75 @@ impl AIR for CairoAIR {
     ) -> BoundaryConstraints<Self::Field> {
-        let last_step = self.context.trace_length - 1;
-
         let initial_pc =
-            BoundaryConstraint::new(MEM_A_TRACE_OFFSET, 0, self.pub_inputs.pc_init.clone());
+            BoundaryConstraint::new(MEM_A_TRACE_OFFSET, 0, public_input.pc_init.clone());
         let initial_ap =
-            BoundaryConstraint::new(MEM_P_TRACE_OFFSET, 0, self.pub_inputs.ap_init.clone());
+            BoundaryConstraint::new(MEM_P_TRACE_OFFSET, 0, public_input.ap_init.clone());
 
         let final_pc = BoundaryConstraint::new(
             MEM_A_TRACE_OFFSET,
-            last_step,
-            self.pub_inputs.pc_final.clone(),
+            self.number_steps - 1,
+            public_input.pc_final.clone(),
         );
         let final_ap = BoundaryConstraint::new(
             MEM_P_TRACE_OFFSET,
-            last_step,
-            self.pub_inputs.ap_final.clone(),
+            self.number_steps - 1,
+            public_input.ap_final.clone(),
         );
 
-        let constraints = vec![initial_pc, initial_ap, final_pc, final_ap];
+        // Auxiliary constraint: permutation argument final value
+        let final_index = self.context.trace_length - 1;
+
+        let mut cumulative_product = FieldElement::one();
+        for (i, value) in public_input.program.iter().enumerate() {
+            cumulative_product = cumulative_product
+                * (&rap_challenges.z_memory
+                    - (FieldElement::from(i as u64 + 1) + &rap_challenges.alpha_memory * value));
+        }
+        let permutation_final =
+            rap_challenges.z_memory.pow(public_input.program.len()) / cumulative_product;
+        let permutation_final_constraint =
+            BoundaryConstraint::new(PERMUTATION_ARGUMENT_COL_3, final_index, permutation_final);
+
+        let one: FieldElement<Self::Field> = FieldElement::one();
+        let range_check_final_constraint =
+            BoundaryConstraint::new(PERMUTATION_ARGUMENT_RANGE_CHECK_COL_3, final_index, one);
+
+        let range_check_min = BoundaryConstraint::new(
+            RANGE_CHECK_COL_1,
+            0,
+            FieldElement::from(public_input.range_check_min.unwrap() as u64),
+        );
+        let range_check_max = BoundaryConstraint::new(
+            RANGE_CHECK_COL_3,
+            final_index,
+            FieldElement::from(public_input.range_check_max.unwrap() as u64),
+        );
+
+        let constraints = vec![
+            initial_pc,
+            initial_ap,
+            final_pc,
+            final_ap,
+            permutation_final_constraint,
+            range_check_final_constraint,
+            range_check_min,
+            range_check_max,
+        ];
 
         BoundaryConstraints::from_constraints(constraints)
     }
 
     fn context(&self) -> AirContext {
         self.context.clone()
     }
+
+    fn composition_poly_degree_bound(&self) -> usize {
+        2 * self.context().trace_length
+    }
```

Comment:
This bound is calculated with the max transition degree and the trace length, right?
If this is the case, this could be defined directly in the AIR trait with a default implementation, taking the max of the degrees stored in the AirContext.

On another note, should't this be the trace lenght - 1? Maybe I am missing something